### PR TITLE
HM refactor

### LINKED ISSUE #406: Refactor: implement history matching as an active learner

### PR DESCRIPTION
**Pull request by radka-j**
Closes #406 

## Overview

This PR:
- adds two new classes in `experimental` based on `HistoryMatching` in main
    - `HistoryMatching` is a metric that takes in observations and returns implausibility for provided predictions
        - it does not require an emulator or a simulator to be instantiated
        - one can optionally pass in an emulator in which case the class provides a method for making predictions with it 
    - `HistoryMatchingWorkflow` captures the iterative process of using an emulator to choose what simulations to run and then updating that emulator with the newly simulated data
        - it expects the refactored `Simulator` and `GaussianProcessExact` emulators as input 
        - each time `run()` is called, it now executes just one "wave", because the method stores state, the user can simply call `run()` repeatedly if they want to run multiple waves 
        - when NROY parameters get identified, the `simulator.param_bounds` get updated to the NROY min/max, this way we can just keep using the simulator `sample_inputs` method without having to have a separate method for sampling
- adds a notebook to `experimental/exploratory/hm_refactor.ipynb` that shows the two HM classes and how they integrate with the dahsboard
- updates the HM Dashboard to accept tensors as inputs 

## For reviewers:
- The added `experimental/exploratory/hm_refactor.ipynb` notebook is a good place to start
- It would be good to double check the workflow implemented in `HistoryMatchingWorkflow.run()`, it is based on what I understood of what is described in this [paper](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1011257), it'd be great to see if we all agree on this.

### CODE REVIEW

File: autoemulate/experimental/history_matching.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+        """
+        Initialize the history matching workflow object.
+
+        Parameters
+        ----------
+        simulator: Simulator
+            A simulator.
+        emulator: GaussianProcessExact
+            NOTE: this can be other GP emulators when implemented.
+            A Gaussian Process emulator pre-trained on `simulator` data.
```
[the remark this answers is not in the document]
Comment (radka-j) [reply autoemulate/experimental/history_matching.py:253]: I think it has to be anything with uncertainty predictions.

File: autoemulate/experimental/history_matching.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        Parameters
+        ----------
+        pred_means: TensorLike
+            Tensor of prediction means [n_samples, n_outputs]
+        pred_vars: TensorLike | None
+            Tensor of prediction variances [n_samples, n_outputs]. If not
+            provided (e.g., when predictions are made by a deterministic
+            simulator), all variances are set to `default_var`.
+        default_var: int
+            TODO: should this just be 0?
```
Comment (radka-j) [inline autoemulate/experimental/history_matching.py:200]: I think having some variance estimate is a better default 
```suggestion

```

**Review by sgreenbury — COMMENTED** [review]
Thanks @radka-j for this PR and discussion - I've added a few comments below! I think the overall structure is looking great and it seems like the main remaining aspect is a change to the run loop to work with implausibility calculations only from the emulator inside the run loop.

File: autoemulate/experimental/history_matching.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+        I_i(\bar{x_0}) = \frac{|z_i - \\mathbb{E}(f_i(\bar{x_0}))|}
+        {\\sqrt{\text{Var}[z_i - \\mathbb{E}(f_i(\bar{x_0}))]}}
+
+    Query points above a given implausibility threshold are ruled out (RO)
+    whereas all other points are marked as not ruled out yet (NROY).
+    """
+
+    def __init__(
+        self,
+        observations: Union[dict[str, tuple[float, float]], dict[str, float]],
```
[the remark this answers is not in the document]
Comment (sgreenbury) [reply autoemulate/experimental/history_matching.py:26]: Opened https://github.com/alan-turing-institute/autoemulate/issues/540

[the remark this answers is not in the document]
Comment (radka-j) [reply autoemulate/experimental/history_matching.py:26]: Thanks!

File: autoemulate/experimental/history_matching.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (radka-j) [reply autoemulate/experimental/history_matching.py:241]: Yes that would be helpful.

File: autoemulate/experimental/history_matching.py [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+        n_simulations: int
+            The maximum number of simulations to run.
+        n_test_samples: int
+            Number of input parameters to test for implausibility with the emulator.
+            Parameters to simulate are sampled from this NROY subset.
+
+        Returns
+        -------
+        tuple[TensorLike, TensorLike]
+            A tensor of tested input parameters and their imaplausability scores from
```
Comment (sgreenbury) [inline autoemulate/experimental/history_matching.py:412]:
```suggestion
            A tensor of tested input parameters and their implausibility scores from
```

File: autoemulate/experimental/history_matching.py [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+        to_simulate_x = self.sample_tensor(n_simulations, nroy_x)
+
+        # Make predictions using simulator (this updates self.x_train and self.y_train)
+        _, _ = self.simulate(to_simulate_x)
+
+        # Refit emulator using all available data
+        self.emulator.fit(self.train_x, self.train_y)
+
+        # Return test parameters and impl scores for this run/wave
+        return test_x, impl_scores
```
Comment (sgreenbury) [inline autoemulate/experimental/history_matching.py:435]: Adding a potential draft suggestion for an additional method wrapping `run()` - wondering if it's helpful to illustrate how to do this/support some use cases where it would not fail? However, as we discussed perhaps the additional complexity of this method handling edge cases (such as where no NROY samples are returned, or the simulator fails) means this could be unhelpful? Also very happy with not implementing this as can also see it over-complicating the aim of the workflow.

```suggestion
        return test_x, impl_scores
        
    def run_exact(self, n_simulations: int=100, n_test_samples: int=10000, max_retries: int = 3) -> tuple[TensorLike, TensorLike]: 
        """Run one or many waves of history matching to run exactly `n_simulations`."""
        test_parameters_list, impl_scores_list = [], []
        retries = 0
        while torch.cat(test_parameters_list, 0).shape[0] < n_simulations:
            test_parameters, impl_scores = self.run(
                n_simulations=n_simulations, n_test_samples=n_test_samples
            )
            if test_parameters.shape[0] == 0:
                retries += 1
            if retries == max_retries:
                msg = (
                    f"Could not generate n_simulations ({n_simulations}) samples "
                    "that are NROY after {max_retries} retries."
                )
                raise RuntimeError(msg)
            test_parameters_list.append(test_parameters)
            impl_scores_list.append(impl_scores)
        return torch.cat(test_parameters_list, 0), torch.cat(impl_scores_list, 0)
```

**Review by sgreenbury — APPROVED** [review]
I think the revised `HistoryMatchingWorkflow` with the new distinct methods and updated `run()` looks really great! The workflow looks like it reflects paper to me (I had one small query regarding the update). I also had one suggestion regarding the potential additional method but I think this looks good to me to merge without that too.

File: autoemulate/experimental/history_matching.py [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+        """
+
+        # Generate `n_test_samples` of parameters and get implausability scores
+        test_x, impl_scores = self.generate_samples(n_test_samples)
+
+        # Rule out implausible parameters from samples
+        nroy_x = self.get_nroy(impl_scores, test_x)
+
+        # Update simulator parameter bounds to NROY -> next time sample from this region
+        self.update_simulator_bounds(nroy_x)
```
[the remark this answers is not in the document]
Comment (radka-j) [reply autoemulate/experimental/history_matching.py:423]: My understanding of the figure is that you restrict the next wave sampling to the NROY space and for implementation purposes. I decided to do that by updating the simulator bounds as that gives us a LHC sampling method. In contrast, the HM implementation in main takes min/max of the NROY samples and does random sampling within those.

I share your worry about restricting the bounds if I don't have enough samples. This is another reason why I'd prefer a user just calls the methods within `runI()` sequentially and inserts checks between them and handles edge cases like that.

But thinking about it now, we should just add some `min_nroy` argument to the function that controls how many samples you want before updating the bounds.

File: autoemulate/experimental/history_matching.py (hunk 5 above)
Comment (radka-j) [reply autoemulate/experimental/history_matching.py:435]: I like the idea of adding a `max_retries` arg but I think it should be an input to `run()`. This is because I think we want to ensure there are `n_simulations` one could run without updating the sampling bounds and refitting the emulator.

So we'd do something like:

```python
def run(..., max_retries = 3):
    test_parameters_list, impl_scores_list, nroy_parameters_list = [], [], []
    
    while torch.cat(nroy_parameters_list, 0).shape[0] < n_simulations:
        test_x, impl_scores = self.generate_samples(n_test_samples)
        nroy_x = self.get_nroy(impl_scores, test_x)
        nroy_parameters_list.append(nroy_x)
        
        retries += 1
        if retries == max_retries:
            msg = (
                f"Could not generate n_simulations ({n_simulations}) samples "
                f"that are NROY after {max_retries} retries."
            )
                raise RuntimeError(msg)
         ...
         
     self.update_simulator_bounds(nroy_x)
        
```

If this fails then the user has an option of increasing the number of test samples to generate or decreasing the number of simulations to run.

File: autoemulate/experimental/history_matching.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (radka-j) [reply autoemulate/experimental/history_matching.py:423]: Of course if we do the update suggested above, then we don't need a `min_nroy` argument as it will be guaranteed to  be at least `n_simulations`.

**Comment by radka-j** [comment]
@sgreenbury this should be ready now but it would be great if you could have one last look over it. Specifically:
- I update the `run()` method, as agreed it now takes in `max_tries` as an input argument. 
- I also made some additional commits trying to ensure device is handled consistently throughout and I'd appreciate if you could have one more look over the whole HM code with just that in mind and see if I missed anything.

**Comment by sgreenbury** [comment]
> @sgreenbury this should be ready now but it would be great if you could have one last look over it. Specifically:
> 
>     * I updated the `run()` method, as agreed it now takes in `max_tries` as an input argument.

Looks great!

>     * I also made some additional commits trying to ensure device is handled consistently throughout and I'd appreciate if you could have one more look over the whole HM code with just that in mind and see if I missed anything.

Good point and looks good! I opened #562 to explore adding a test for `run()` and adding one `cpu()` call and a dtype in the subclassed `Epidemic` simulator. I am not sure if we need `Simulator` to consider `Device` at the moment? More generally, I am wondering how we could do testing across devices more consistently across PRs (without adding lots of boilerplate/extra tests) - might be worth looking at in a new issue. Also possibly relates to potential integration with lightning.

My only other minor thought was about the additional handling on the `None` case for the emulator in `HistoryMatching` and whether the emulator is needed in `HistoryMatching` or if it could be moved to the `HistoryMatchingWorkflow`?

**Comment by radka-j** [comment]
@sgreenbury 
- Thank you for adding the test!
- I added an issue for adding device handling to Simulator, it's been on my mind since starting to work with TorchCor which is implemented in PyTorch. 
- As for the emulator, I originally didn't have that option in the main class but added it because it makes getting predictions from the emulator in the right format for scoring implausibility easier for the user. It's demonstrated in the notebook. Alternatively it could be a method of the emulator to get just the mean and the variance like this?

**Review by sgreenbury — APPROVED** [review]
